### DISCUSSION TITLE
Question about imperative vs functional

### DISCUSSION THREAD

**Discussion by bmadone** [gcanti/fp-ts#1797] [Q&A]
Hi, I started using fp-ts for my personal project, and don't understand where is this border between functional and imperative code, and how to create these borders. For example for data fetching I use react-query, I don't see how I can combine it with functional way like TaskEither and etc. Also I validate api data with zod, it's just enough for me, it's solid library with great support, what is the purpose of io-ts, is it really better? And what about null/undefined, should I reduce them completely, or it's normal to have them together with Option?

For example I use Option if I have a pipe after it, and the same for array operations, if I don't need pipe, I use native methods in array, it is wrong, or not?

**Comment by JoshuaAmaju** [comment]
I do use fp-ts with React Query, and I looooove it
